src/window/event_time_clock.rs

```diff
@@ -0,0 +1,276 @@
+use std::task::Poll;
+
+use pyo3::{exceptions::PyValueError, pyclass, pymethods, PyAny, PyResult, Python, ToPyObject};
+
+use crate::{
+    pyo3_extensions::{chrono_tz::ChronoDateTime, TdPyCallable, TdPyAny},
+    recovery::StateBytes,
+    window::ClockConfig,
+};
+
+use super::{
+    system_clock::SystemClock,
+    testing_clock::{TestingClock, TestingClockConfig},
+    Builder, Clock, ClockBuilder,
+};
+
+/// Use datetimes from events as clock.
+///
+/// If the dataflow has no more input, all windows are closed.
+///
+/// The watermark is the system time since the last element
+/// plus the value of `late`. It is updated every time an event
+/// with a newer datetime is processed.
+///
+/// Args:
+///   getter: Python function to get a datetime from an event.
+///   late_after_system_duration: How much (system) time to wait before considering an event late.
+///   system_clock_config: TODO
+///
+/// Returns:
+///   Config object. Pass this as the `clock_config` parameter to your
+///   windowing operator.
+#[pyclass(module="bytewax.window", extends=ClockConfig)]
+#[pyo3(text_signature = "(getter, late_after_system_duration, system_clock_config)")]
+#[derive(Clone)]
+pub(crate) struct EventClockConfig {
+    #[pyo3(get)]
+    pub(crate) dt_getter: TdPyCallable,
+    #[pyo3(get)]
+    pub(crate) late_after_system_duration: chrono::Duration,
+    #[pyo3(get)]
+    pub(crate) system_clock_config: Option<TestingClockConfig>,
```

Comment:
We have an optional TestingClockConfig here to allow for testing windowing with event time?

Reply:
It's now slightly different, but yes, we still accept a parameter to use a testing clock rather than the default system clock in the event time clock. As I said in another comment, initially I implemented two separate clocks, `EventTimeClock` and `TestingEventTimeClock`, with the only difference between the two that the testing clock was using an auto incremented counter rather than system's `now`.
This meant keeping the two implementations in sync by hand, and could lead to problems if not (a test passing when it shouldn't because test and production use two different implementation of the event time clock), so I decided to find a way to reuse the same EventClock changing only the meaning of `now`.
I'm still not convinced it's the best approach, since there is a parameter needed only for testing inside a "production" object, but I felt this was better than the way I did it previously.

Open to suggestions though